tree/tree/src/TTree.cxx

```diff
@@ -5492,7 +5492,7 @@ void TTree::InitializeBranchLists(bool checkLeafCount)
    }
 
    // The special branch fBranchRef also needs to be processed sequentially
-   if (fBranchRef) {
+   if (fBranchRef && fBranches.FindObject(fBranchRef)) {
```

Comment:
I don't understand the semantic/meaning of this additional test.  Why wouldn't fBranchRef be sequentially processed if it is not also in fBranches?  Also in which case is it in fBranches, in which case is it not?

Reply:
I just updated the commend under 
https://sft.its.cern.ch/jira/browse/ROOT-8995.

Reply:
In which case, the patch is in the wrong direction.  fBranchRef must be processed first (and thus sequentially) but (if I read correctly the code you point to) must be filtered out of the normal list of branches (and/or their scheduling).

Reply:
In MainEvent.cxx, it create a single branch "Event" and fBranchRef is a subbranch of "Event". This line is adding fBranchRef in fSeqBranches. I guess it is onlyt supposed to add "Event".

Reply:
Why?

Reply:
I edited my previous comment since there was some delay.

Reply:
> In this case, "Event" and all its subbranches should be processed in sequential or parallel?

The branch refered by fBranchRef should be first and sequential.  All the other branch should be (once TBranchRef is loaded) read in parallel.

Reply:
In other words,

The decompression in above case should go to

https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5370

or 

https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5416

Reply:
In current ROOT code, 
https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5495
adds TRefTable to fSeqBranches and this is the only branch in fSeqBranches before the code is going to 
https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5370
and the above line processes TRefTable instead of Event.
Consequently, 
https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5416
nbranches is 1(Event) and fSeqBranches.size() is also 1(fBranchRef) and TThreadExecutor is doing nothing here.

Reply:
@pcanal 
When I run 
./test/eventexe 5000 6 99 21 1000 1
to unzip baskets in parallel. What should be the correct order? What do you mean by loading TRefTable(I don't think it should processed by TBranch::GetEntry(), correct?).

Reply:
> https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5416
> nbranches is 1(Event) and fSeqBranches.size() is also 1(fBranchRef) 

That logic is wrong ....

Reply:
I have been confused by
https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5416

What is the correct logic?

Reply:
> What should be the correct logic?

Not sure.  One need to review the whole routine.  I suspect one has to just skip the branch that are in SeqBranches during the loop.

Reply:
As Philippe points out, the issue is in:
https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5416

In that line, if fBranchRef is not null (and thus it has been added to fSeqBranches), we need to take that into account and modify the "nbranches - fSeqBranches.size()" substraction.

Something like this:
```
Int_t numParBanches = nbranches - fSeqBranches.size();
if (fBranchRef) ++numParBranches;
pool.Foreach(mapFunction, numParBranches);
```

Reply:
Or even simpler:
```
pool.Foreach(mapFunction, fSortedBranches.size());
```

Reply:
What about the sequential processing on fSeqBranches in the following line?

https://github.com/root-project/root/blob/master/tree/tree/src/TTree.cxx#L5370

In my case, fBranchRef is the only branch in fSeqBranches, would calling GetEntry for fBranchRef be the correct behavior as you expect?

Reply:
In your case there is only fBranchRef as sequential branch, but in the general case there could be others, or even fBranchRef could be null. The sequential branches are normally leaf count branches that store the size of some other branch, and such branches need to be processed sequentially first in the loop that you pointed out.

I pushed the fix to master and 6.10 patches, could you confirm that it solves your issue?

Reply:
I tested latest master. It worked fine. Thanks!